### DISCUSSION TITLE
Failed with: Unable to find migration version to revert in the migrations directory.

### DISCUSSION THREAD

**Discussion by Yoric** [diesel-rs/diesel#2719] [Q&A]
Diesel used to work nicely.
All of a sudden, after a `git pull --rebase` (I think), 

```sh
$ diesel migration revert
Failed with: Unable to find migration version to revert in the migrations directory.
```

However,

```sh
$ diesel migration list
Migrations:
  [X] 00000000000000_diesel_initial_setup
  [X] 2021-03-25-160646_rooms_to_ignore
  [X] 2021-03-30-160520_keywords
  [X] 2021-04-09-121228_descriptions
```

Does anyone know what is going on and what I can do to fix it?

I'm running diesel_cli v1.4.1 on postgresql.

**Comment by weiznich** [comment]
This is expected behavior, thought the error message could be clearer here.
Diesel tracks which migrations are applied and which are not applied yet in a database table. This table does not contain the migration itself or the corresponding down migration. To revert a migration the corresponding migration is searched in your local migration folder based on the migration version (that's by default the timestamp in front of your migration name). If no migration with this version is found locally this error message is returned.

That all written: Improving this error message is definitively possible. If someone is interested in contributing here feel free to get in touch with me to get some more specific information about were to look for the corresponding code.

**Reply by Yoric** [reply]
That's basically what I had understood. However, I still don't get why it's not finding the migration or what I need to do to solve my problem.

Right now, I've run a `diesel database reset`, which is acceptable because it happened on a dev machine (I lost data, but I can live with it), but if it happens in production, I have no idea how to not hit a brick wall.

**Reply by weiznich** [reply]
It's not finding the migration because it's not there on your file system. When you execute `diesel migration revert` the cli tool looks up the version of the last applied migration in a internal database table. After that it looks for a migration with that version in your migration folder (either that one you supplied explicitly via `--migration-dir` or the default `migrations` folder). If you now execute a `git rebase` command (or any other command) that changes the content of your migration folder it can miss that specific migration.

That written: You should likely never execute a `diesel migration revert` on a production database. Instead it's likely a much better idea to use the `down.sql` file of a specific migration to roll out a new migration incrementing the migration version further.

**Reply by Yoric** [reply]
Since `diesel migration list` finds the migrations and their current state, shouldn't `diesel migration revert` be able to find them, too?

Perhaps with an arg `--force`?

**Reply by weiznich** [reply]
Again: The migration you are trying to revert does not exist in your local file system, but diesel cli expect it to exist, therefore the error.
`diesel migration list` only shows the state of migrations that exist in your local file system. So if your database contains another migration that does not exist it's not shown there. I do not see how an argument like `--force` would help here. That would basically mean that we just remove the corresponding entry from our internal migrations table and otherwise leave the database unchanged. I'm quite sure that's not the behaviour anyone would expect there.

**Reply by Yoric** [reply]
Apparently, there is something that diesel cli does and that I do not understand.

As far as I understand, the sole state of diesel cli is `__diesel_schema_migrations`, which only contains the list of migrations it knows and whether they have been run. My understanding is that both `diesel migration list` and `diesel migration revert` look at this table to determine which migrations have been run/need to be reverted, right?

Are you telling me that in my case, one of the migrations was in `__diesel_schema_migrations` but not in my `migrations/` directory? I don't *think* that this is what happened but I may be wrong.

If that's the case, yes, a better error message that mentions which migration is missing (or at least its version) would be really useful.

**Reply by weiznich** [reply]
> Are you telling me that in my case, one of the migrations was in __diesel_schema_migrations but not in my migrations/ directory? I don't think that this is what happened but I may be wrong.

That's exactly your issue.

> If that's the case, yes, a better error message that mentions which migration is missing (or at least its version) would be really useful.

We are happy to receive a contribution here. Otherwise this is possibly already fixed by https://github.com/diesel-rs/diesel/pull/2657

**Reply by Yoric** [reply]
PR opened: https://github.com/diesel-rs/diesel/pull/2723 .

It should help a bit.

**Reply by Mingun** [reply]
While design goals are understandable here, I think, that diesel CLI could handle this situation better:
- `diesel migration list` could check the filesystem _and_ the database and output migrations from both lists merged. The migrations that is not in both lists should be marked. The output might looks like:
   ```
   $ diesel migration list
   Migrations (/path/to/the/migration/folder):
     [X] 00000000000000_diesel_initial_setup
     [X] 2021-03-25-160646_rooms_to_ignore (missed file)
     [X] 2021-03-30-160520_keywords (migration file is not tracked)
     [X] 2021-04-09-121228_descriptions

   error: your migrations are corrupted (the exact message could be better, suggesting the ways to resolve this case)
   ```
- `diesel migration revert` can have a `--force/-f` flag to explicitly ignore absence of missed files / DB entries and assume that nothing to do in that case
- Diesel can provide a new command `diesel migration repair` which will recreate internal DB state from existing files in migration directory, maybe asking if file should be applied/marked as applied/marked as unapplied/skipped and deleted

That's my 2 cents. It's all about quality of life issues which distinguish simply good software from excellent.

**Reply by weiznich** [reply]
I agree that this could be handled in a better way, but after all that would need someone that thinks that this is more important than other features and spend time to:

* bring up a complete design
* implement the changes
* adds tests + documentation for these changes.

There is unfortunately only that much time everyday, so I won't have the capacity for that. Feel free to contribute such changes.
